crates/loans/src/farming.rs

```diff
@@ -0,0 +1,200 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use sp_io::hashing::blake2_256;
+use sp_runtime::{traits::Zero, DispatchResult};
+
+use crate::*;
+
+impl<T: Config> Pallet<T> {
+    pub(crate) fn reward_account_id() -> Result<T::AccountId, DispatchError> {
+        let account_id: T::AccountId = T::PalletId::get().into_account_truncating();
+        let entropy = (b"loans/farming", &[account_id]).using_encoded(blake2_256);
+        Ok(T::AccountId::decode(&mut &entropy[..]).map_err(|_| Error::<T>::CodecError)?)
+    }
+
+    fn reward_scale() -> u128 {
+        10_u128.pow(12)
+    }
+
+    fn calculate_reward_delta_index(
+        delta_block: T::BlockNumber,
+        reward_speed: BalanceOf<T>,
+        total_share: BalanceOf<T>,
+    ) -> Result<u128, sp_runtime::DispatchError> {
+        if total_share.is_zero() {
+            return Ok(0);
+        }
+        let delta_block: BalanceOf<T> = delta_block.saturated_into();
+        let delta_index = delta_block
+            .get_big_uint()
+            .checked_mul(&reward_speed.get_big_uint())
+            .and_then(|r| r.checked_mul(&Self::reward_scale().get_big_uint()))
+            .and_then(|r| r.checked_div(&total_share.get_big_uint()))
+            .and_then(|r| r.to_u128())
+            .ok_or(ArithmeticError::Overflow)?;
+        Ok(delta_index)
+    }
+
+    fn calculate_reward_delta(
+        share: BalanceOf<T>,
+        reward_delta_index: u128,
+    ) -> Result<u128, sp_runtime::DispatchError> {
+        let reward_delta = share
+            .get_big_uint()
+            .checked_mul(&reward_delta_index.get_big_uint())
+            .and_then(|r| r.checked_div(&Self::reward_scale().get_big_uint()))
+            .and_then(|r| r.to_u128())
+            .ok_or(ArithmeticError::Overflow)?;
+        Ok(reward_delta)
+    }
+
+    pub(crate) fn update_reward_supply_index(asset_id: AssetIdOf<T>) -> DispatchResult {
+        let current_block_number = <frame_system::Pallet<T>>::block_number();
+        RewardSupplyState::<T>::try_mutate(asset_id, |supply_state| -> DispatchResult {
+            let delta_block = current_block_number.saturating_sub(supply_state.block);
+            if delta_block.is_zero() {
+                return Ok(());
+            }
+            let supply_speed = RewardSupplySpeed::<T>::get(asset_id);
+            if !supply_speed.is_zero() {
+                let total_supply = Self::total_supply(asset_id)?;
+                let delta_index = Self::calculate_reward_delta_index(delta_block, supply_speed, total_supply)?;
+                supply_state.index = supply_state
+                    .index
+                    .checked_add(delta_index)
+                    .ok_or(ArithmeticError::Overflow)?;
+            }
+            supply_state.block = current_block_number;
+
+            Ok(())
+        })
+    }
+
+    pub(crate) fn update_reward_borrow_index(asset_id: AssetIdOf<T>) -> DispatchResult {
```

Comment:
I think in general we could do with more documentation, functions like this are a little bit of a black box to me

Reply:
I agree, the original code wasn't documented and I prioritised feature work. I don't think this should block the testnet release, created this ticket: https://github.com/interlay/interbtc/issues/756